baselines/fedper/README.md

```diff
@@ -0,0 +1,158 @@
+In this Baseline (FedPer), we must store the state of the local client head while aggregation of body parameters happen at the server. Flower is currently making this possible but for the time being, we reside to storing client _head_ state in a folder called client_states. We store the values after each fit and evaluate function carried out on each client, and call for the state before executing these funcitons. Moreover, the state of a unique client is accessed using the client ID. 
+
+> NOTE: This is a work-around so that the local head parameters are not reset before each fit and evaluate. Nevertheless, it can come to change with future releases. 
+
+
+## Environment Setup
+
+To construct the Python environment follow these steps:
+
+```bash
+# Set Python 3.10
+pyenv local 3.10.6
+# Tell poetry to use python 3.10
+poetry env use 3.10.6
+
+# install the base Poetry environment
+poetry install
+
+# activate the environment
+poetry shell
+```
+## Running the Experiments
+```bash
+python -m fedper.main # this will run using the default settings in the `conf/base.yaml` 
+
+# When running models for flickr dataset, it is important to keep batch size at 4 or lower since some clients (for reproducing experiment) will have very few examples of one class
+```
+
+To reproduce figures:
+```bash
+# make fedper/run_scripts.sh executable
+chmod u+x fedper/run_scripts.sh
+# uncomment lines in script that you want to run, then  
+bash fedper/run_scripts.sh
+
+# this config can also be overriden from the CLI
+```
+
+What you need to change in configuration files: 
+```bash
+algorithm: fedavg, fedper # these are currently supported
+server_device: 'cuda:0', 'cpu'
+dataset.name: 'cifar10', 'flickr'
+num_classes: 10, 100, 5 # respectively 
+dataset.num_classes: 4, 8, 10 #for non-iid split assigning n num_classes to each client (these numbers for CIFAR10 experiments)
+model.name: mobile, resnet
+```
+
+To run multiple runs, one can also reside to `HYDRA`'s multirun option. 
+```bash
+# for CIFAR10
+python -m fedper.main --multirun dataset.num_classes=4,8,10 model.name=resnet,mobile algorithm=fedper,fedavg model.num_head_layers=2,3
+
+# to repeat each run 5 times, one can also add
+python -m fedper.main --multirun dataset.num_classes=4,8,10 model.name=resnet,mobile algorithm=fedper,fedavg model.num_head_layers=2,3 '+repeat_num=range(5)'
+```
+
+
+## Expected Results
+
+Having run the `run_script.sh`, the expected results should look something like this: 
+
+**MobileNet-v1 on CIFAR10**
+
+![](_static/mobile_plot_figure_2.png)
+
+**ResNet on CIFAR10**
+
+![](_static/resnet_plot_figure_2.png)
+
+**MobileNet-v1 on CIFAR10 using varying size of head**
+
+![](_static/mobile_plot_figure_num_head.png)
+
+
+**ResNet on CIFAR10 using varying size of head**
+
+![](_static/resnet_plot_figure_num_head.png)
+
+**MobileNet-v1 on FLICKR-AES**
+
+![](_static/mobile_plot_figure_flickr.png)
+
+**ResNet on FLICKR-AES**
+
+![](_static/resnet_plot_figure_flickr.png)
```

Comment:
It's great having these many plots! What do you think about rearranging them in such a way that each experiment for mobilenet and resnet34 is shown  next to each other (instead of above/under). This code will do just that:

```
**MobileNet-v1 and ResNet-34 on CIFAR10**

<img src="_static/mobile_plot_figure_2.png" width="500"/> <img src="_static/resnet_plot_figure_2.png" width="500"/> 

**MobileNet-v1 and ResNet-34 on CIFAR10 using varying size of head**

<img src="_static/mobile_plot_figure_num_head.png" width="500"/> <img src="_static/resnet_plot_figure_num_head.png" width="500"/> 

**MobileNet-v1 and ResNet-34 on FLICKR-AES**

<img src="_static/mobile_plot_figure_flickr.png" width="500"/> <img src="_static/resnet_plot_figure_flickr.png" width="500"/> 
```